File: frontend-learner-dashboard-app/android/app/src/main/java/com/template/app/offlinemedia/OfflineMediaPlugin.java
package io.vacademy.student.app.offlinemedia;

import android.os.StatFs;
import android.util.Base64;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;
import java.io.IOException;

/**
 * Native `OfflineMedia` Capacitor plugin (Android). Local, in-repo plugin — no npm package —
 * so it is not auto-registered via `capacitor.plugins.json`; registered explicitly by
 * `MainActivity.onCreate()` via `registerPlugin(OfflineMediaPlugin.class)`.
 *
 * Provides:
 *  - `getFreeDiskSpace()` — bytes available on the volume backing the app's private files dir.
 *  - `openAsset({path, keyB64, nonceB64, mimeType})` — registers a decrypt session for an
 *    on-disk ciphertext file and returns `{token, url}`. `url` is
 *    `http://127.0.0.1:<port>/<token>/stream`, served by {@link OfflineMediaServer} (a tiny
 *    localhost-only HTTP responder — see that class's doc comment for why a localhost server was
 *    chosen over trying to hook `shouldInterceptRequest`), directly usable as an HTML5
 *    `<video>` src with Range/206 support for seeking.
 *  - `closeAsset({token})` — drops the session and zeroizes its key material.
 */
@CapacitorPlugin(name = "OfflineMedia")
public class OfflineMediaPlugin extends Plugin {

    @PluginMethod
    public void getFreeDiskSpace(PluginCall call) {
        try {
            StatFs stat = new StatFs(getContext().getFilesDir().getPath());
            long bytes = stat.getAvailableBytes();
            JSObject ret = new JSObject();
            ret.put("bytes", bytes);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to read free disk space: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void openAsset(PluginCall call) {
        String path = call.getString("path");
        String keyB64 = call.getString("keyB64");
        String nonceB64 = call.getString("nonceB64");
        String mimeType = call.getString("mimeType");

        if (path == null) {
            call.reject("Missing required parameter: path");
            return;
        }
        if (keyB64 == null) {
            call.reject("Missing required parameter: keyB64");
            return;
        }
        if (nonceB64 == null) {
            call.reject("Missing required parameter: nonceB64");
            return;
        }

        byte[] key;
        byte[] nonce;
        try {
            key = Base64.decode(keyB64, Base64.NO_WRAP);
            nonce = Base64.decode(nonceB64, Base64.NO_WRAP);
        } catch (IllegalArgumentException e) {
            call.reject("keyB64/nonceB64 must be valid base64: " + e.getMessage());
            return;
        }

        if (key.length != 32) {
            call.reject("keyB64 must decode to 32 bytes (AES-256), got " + key.length);
            return;
        }
        if (nonce.length != 16) {
            call.reject("nonceB64 must decode to 16 bytes, got " + nonce.length);
            return;
        }

        File file = new File(path);
        if (!file.exists()) {
            call.reject("File does not exist at path: " + path);
            return;
        }

        String resolvedMimeType = mimeType != null ? mimeType : guessMimeType(path);

        try {
            int port = OfflineMediaServer.INSTANCE.ensureStarted();
            String token = OfflineMediaSessionStore.INSTANCE.open(path, key, nonce, resolvedMimeType);

            JSObject ret = new JSObject();
            ret.put("token", token);
            ret.put("url", "http://127.0.0.1:" + port + "/" + token + "/stream");
            call.resolve(ret);
        } catch (IOException e) {
            call.reject("Failed to start OfflineMedia localhost server: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void closeAsset(PluginCall call) {
        String token = call.getString("token");
        if (token == null) {
            call.reject("Missing required parameter: token");
            return;
        }
        OfflineMediaSessionStore.INSTANCE.close(token);
        call.resolve();
    }

    private static String guessMimeType(String path) {
        String lower = path.toLowerCase();
        if (lower.endsWith(".webm")) return "video/webm";
        return "video/mp4";
    }
}

File: frontend-learner-dashboard-app/ios/App/App/OfflineMediaPlugin.swift
import Foundation
import Capacitor

/// One open decrypt "session". Key material lives only in memory for the lifetime of the
/// session and is zeroized when the session is closed (see `OfflineMediaSessionStore.close`).
final class OfflineMediaSession {
    let path: String
    var key: [UInt8]
    var nonce: [UInt8]
    let mimeType: String

    init(path: String, key: [UInt8], nonce: [UInt8], mimeType: String) {
        self.path = path
        self.key = key
        self.nonce = nonce
        self.mimeType = mimeType
    }

    /// Overwrites key/nonce bytes with zeros before the session is dropped, so no key
    /// material lingers in the process heap longer than necessary.
    func zeroize() {
        for i in 0..<key.count { key[i] = 0 }
        for i in 0..<nonce.count { nonce[i] = 0 }
    }
}

/// In-memory registry of open `OfflineMedia.openAsset()` sessions, keyed by opaque token.
/// Shared between `OfflineMediaPlugin` (which creates/destroys sessions) and
/// `OfflineMediaSchemeHandler` (which reads from them while serving `offline-media://` requests).
final class OfflineMediaSessionStore {
    static let shared = OfflineMediaSessionStore()

    private let lock = NSLock()
    private var sessions: [String: OfflineMediaSession] = [:]

    private init() {}

    func open(path: String, key: [UInt8], nonce: [UInt8], mimeType: String) -> String {
        let token = UUID().uuidString
        lock.lock()
        sessions[token] = OfflineMediaSession(path: path, key: key, nonce: nonce, mimeType: mimeType)
        lock.unlock()
        return token
    }

    func session(for token: String) -> OfflineMediaSession? {
        lock.lock()
        defer { lock.unlock() }
        return sessions[token]
    }

    func close(token: String) {
        lock.lock()
        let session = sessions.removeValue(forKey: token)
        lock.unlock()
        session?.zeroize()
    }
}

/// Native `OfflineMedia` Capacitor plugin.
///
/// Provides:
///  - `getFreeDiskSpace()` — bytes available on the volume backing the app's home directory.
///  - `openAsset({path, keyB64, nonceB64, mimeType})` — registers a decrypt session for an
///    on-disk ciphertext file and returns a `{token, url}` pair. `url` is an
///    `offline-media://<token>/stream` URL, servable by `OfflineMediaSchemeHandler`
///    (registered on the WKWebView's configuration in `MainViewController`) directly as an
///    HTML5 `<video>` src, including Range/206 support for seeking.
///  - `closeAsset({token})` — drops the session and zeroizes its key material.
@objc(OfflineMediaPlugin)
public class OfflineMediaPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "OfflineMediaPlugin"
    public let jsName = "OfflineMedia"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getFreeDiskSpace", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openAsset", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "closeAsset", returnType: CAPPluginReturnPromise)
    ]

    /// URL scheme used for streaming decrypted video; must match the scheme registered via
    /// `webViewConfiguration.setURLSchemeHandler` in `MainViewController.swift`.
    static let scheme = "offline-media"

    @objc func getFreeDiskSpace(_ call: CAPPluginCall) {
        do {
            let homeURL = URL(fileURLWithPath: NSHomeDirectory())
            let values = try homeURL.resourceValues(forKeys: [.volumeAvailableCapacityForImportantUsageKey])
            let bytes = values.volumeAvailableCapacityForImportantUsage ?? 0
            call.resolve(["bytes": bytes])
        } catch {
            call.reject("Failed to read free disk space: \(error.localizedDescription)")
        }
    }

    @objc func openAsset(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            return call.reject("Missing required parameter: path")
        }
        guard let keyB64 = call.getString("keyB64"), let keyData = Data(base64Encoded: keyB64) else {
            return call.reject("Missing or invalid required parameter: keyB64")
        }
        guard let nonceB64 = call.getString("nonceB64"), let nonceData = Data(base64Encoded: nonceB64) else {
            return call.reject("Missing or invalid required parameter: nonceB64")
        }
        guard keyData.count == 32 else {
            return call.reject("keyB64 must decode to 32 bytes (AES-256), got \(keyData.count)")
        }
        guard nonceData.count == 16 else {
            return call.reject("nonceB64 must decode to 16 bytes, got \(nonceData.count)")
        }
        guard FileManager.default.fileExists(atPath: path) else {
            return call.reject("File does not exist at path: \(path)")
        }

        let mimeType = call.getString("mimeType") ?? OfflineMediaPlugin.guessMimeType(path: path)

        let token = OfflineMediaSessionStore.shared.open(
            path: path,
            key: [UInt8](keyData),
            nonce: [UInt8](nonceData),
            mimeType: mimeType
        )
        let url = "\(OfflineMediaPlugin.scheme)://\(token)/stream"
        call.resolve(["token": token, "url": url])
    }

    @objc func closeAsset(_ call: CAPPluginCall) {
        guard let token = call.getString("token") else {
            return call.reject("Missing required parameter: token")
        }
        OfflineMediaSessionStore.shared.close(token: token)
        call.resolve()
    }

    static func guessMimeType(path: String) -> String {
        let ext = (path as NSString).pathExtension.lowercased()
        switch ext {
        case "m4v", "mov": return "video/mp4"
        case "webm": return "video/webm"
        default: return "video/mp4"
        }
    }
}
